enhancements/builds/node-cache-for-builds.md

```diff
@@ -0,0 +1,138 @@
+  And builds run faster.
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. The build controller will mount the node image cache located at
+   `/var/lib/containers/storage` into the build pod as `readOnly`. The mount point should not
+   conflict with buildah's own default storage directory. Ex: `/var/lib/shared/containers/storage`.
+2. The openshift/builder image is modified to configure additional image stores for buildah (via
+   `/etc/containers/storage.conf`).
+3. When using the additional image store, buildah _copies_ any blobs found in the cache to its own
+   image store within the container.
+
+### Risks and Mitigations
+
+#### Image layer pruned by kubelet GC
+
+*Risk:* Kubelet removing image layer via GC process
+
+*Mitigation:* Buildah will be responsible for copying layers found in the additional image store
+into its own working container storage within the running build container. This ensures that any
+layers pruned from underneath the container during the build will not break the running build.
+
+#### Builds can modify content in the node's container storage
+
+*Risk:* Writing to the node's container storage is a security risk, can lead to the kubelet running
+rogue images.
+
+*Mitigation:* Node container storage will be mounted as read-only as additional read-only storage.
+
+## Design Details
+
+### Test Plan
+
+1. Configure a single-node machine set with a unique label (ex: `build-test: true`)
+2. Use `oc new-app` to create a sample NodeJS application.
+3. Alter the generated `BuildConfig` and `DeploymentConfig` such that the build and node both
+   run on the test node.
+4. Wait for the deployment to be rolled out onto the test node.
+5. Run `oc start-build` for the `BuildConfig` above. Verify the following:
+   1. The build completes.
+   2. Image layers are pulled from the node cache. Most, if not all, base layers should be cached.
+
+### Graduation Criteria
+
+This will be released directly to GA.
+
+### Upgrade / Downgrade Strategy
+
+The standard upgrade procedure will ensure that the build controller and builder image are rolled
+out. Builds launched after upgrade will be able to take advantage of the cache feature.
+
+On downgrade builds will resume their current behavior and will not have an image cache.
+
+### Version Skew Strategy
+
+The build controller uses leader election to ensure only one controller is creating new builds.
+Once a build has been started, its pod configuration cannot change.
+
+## Implementation History
+
+2020-02-18: Initial draft
+
+## Drawbacks
+
+1. Mounting content from the node requires the build pod to continue running as privileged.
+2. No API changes proposed - there is no means to opt-out of this feature.
+
+## Alternatives
+
+1. Use persistent volumes to cache layers for builds
+   1. General build volume support is being considered as a separate enhancement.
+   2. Automatic provisioning of block storage is not guaranteed in all environments.
```

Comment:
i think this also has challenges in terms of write-atomicity and performance.

plus requiring something additional to GC this new cache.

Reply:
Noted